### ISSUE TITLE
Defining 'pytest_plugins' in a non-top-level conftest is no longer supported

### ISSUE THREAD

**Issue by tekumara** [dbt-labs/dbt-utils#627]
### Describe the bug
When running `pytest` in a repo that contains dbt_utils:

```
================================================================================= ERRORS ==================================================================================
______________________________________________________________________ ERROR collecting test session ______________________________________________________________________
Defining 'pytest_plugins' in a non-top-level conftest is no longer supported:
It affects the entire test suite instead of just below the conftest as expected.
  /Users/tekumara/code/jaffleshop/dbt_packages/dbt_utils/tests/conftest.py
Please move it to a top level conftest file at the rootdir:
  None
For more information, visit:
  https://docs.pytest.org/en/stable/deprecations.html#pytest-plugins-in-non-top-level-conftest-files
========================================================================= short test summary info =========================================================================
ERROR 
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================ 1 error in 0.23s =============================================================================
```

### Steps to reproduce

Create a dbt repo with dbt_utils and run `pytest` eg:

```shell
git clone https://github.com/dbt-labs/jaffle_shop
cd jaffle_shop
cat <<'EOF' > packages.yml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.8.6
EOF
dbt deps
pip install pytest
pytest
```

**Comment by tekumara** [comment]
Our current workaround is to add this to _pyproject.toml_:
```
[tool.pytest.ini_options]
addopts = "--ignore=dbt_packages"
```

Until dbt_utils is fixed.

**Comment by joellabes** [comment]
@dbeatty10 I assume this comes from your adding pytest to the project? Could you have a gander for me?

**Comment by dbeatty10** [comment]
@tekumara thank you for the excellent write-up!

## Short story

I agree with your approach to work around this error message with the following (for pytest >= 6.0).

[`pyproject.toml`](https://docs.pytest.org/en/6.2.x/customize.html#pyproject-toml):
```
# pyproject.toml
[tool.pytest.ini_options]
addopts = "--ignore=dbt_packages"
```

or [`pytest.ini`](https://docs.pytest.org/en/6.2.x/customize.html#pytest-ini):
```
# pytest.ini
[pytest]
addopts = "--ignore=dbt_packages"
```

or [`tox.ini`](https://docs.pytest.org/en/6.2.x/customize.html#tox-ini):
```
# tox.ini
[pytest]
addopts = "--ignore=dbt_packages"
```

or [`setup.cfg`](https://docs.pytest.org/en/6.2.x/customize.html#setup-cfg):
```
# setup.cfg
[tool:pytest]
addopts = "--ignore=dbt_packages"
```

## Longer story

As you noted, there are a few key interwoven interactions happening between:
1. how `pytest` handles subdirectories
2. how `dbt deps` installs dbt packages by unzipping the full contents of the repo to ([defaulting to `dbt_packages`](https://docs.getdbt.com/reference/project-configs/packages-install-path))
3. how dbt_utils pip installs [`dbt-tests-adapter`](https://pypi.org/project/dbt-tests-adapter/), specifies it as a plugin [here](https://github.com/dbt-labs/dbt-utils/blob/7f83115f1c3ab2d3cf6b39605ca3c5471abaab13/tests/conftest.py#L4), and executes it like [this](https://github.com/dbt-labs/dbt-utils/blob/7f83115f1c3ab2d3cf6b39605ca3c5471abaab13/run_functional_test.sh#L3).

This creates the situation where the pytest `pytest_plugins` configuration is supported _until_ it appears in a subdirectory.

`pytest` developer [@nicoddemus](https://twitter.com/nicoddemus) has a succinct explanation for why `pytest` decided to do this here:
https://twitter.com/hackebrot/status/1144951693009981441

<img width="440" alt="image" src="https://user-images.githubusercontent.com/44704949/185167561-9b625053-bc60-410a-b7db-9f12c5285f6b.png">

## Potential paths forward

**TLDR**: Definitely plan on using the workaround you mentioned for the foreseeable future!

It would take further research to determine if any of the following would work or not:
1. Configure CI for dbt_utils to use the [`PYTEST_PLUGINS` environment variable](https://docs.pytest.org/en/7.1.x/reference/reference.html#envvar-PYTEST_PLUGINS) instead of `pytest_plugins` within `conftest.py` (and update all applicable documentation as-needed).
2. Modify [dbt-tests-adapter](https://pypi.org/project/dbt-tests-adapter/) [here](https://github.com/dbt-labs/dbt-core/tree/main/tests/adapter) to be a [pip installable plugin](https://docs.pytest.org/en/7.1.x/how-to/writing_plugins.html#pip-installable-plugins)

**Comment by Trymzet** [comment]
Also commenting out this line works:
```python
pytest_plugins = ["dbt.tests.fixtures.project"]
```
(in `dbt_utils/tests/conftest.pt`)

**Comment by mchesler613** [comment]
I have a lower level directory that needs to use a plugins defined in a top-level conftest.py, `pytest_plugins` variable.  pytest `7.4.4` prevents me from redefining the pytest_plugins in the lower level conftest.py.  How do I call pytest on a lower level directory that uses the plugins defined in the top-level directory's conftest.py file?  Thank you.

**Comment by mchesler613** [comment]
Nevermind, I got this to work by telling pytest to ignore the lower-level conftest.py file in the top level `pytest.ini`, which also defines the `pytest_plugins` variable.

**Comment by binhtran04** [comment]
> Nevermind, I got this to work by telling pytest to ignore the lower-level conftest.py file in the top level `pytest.ini`, which also defines the `pytest_plugins` variable.

@mchesler613 How did you do it specifically? Can you provide some code snippets, please?
